File: PracticeChinese/Module/BeginnerCourse/views/PlayerProgressView/ToggleButton.m
//
//  ToggleButton.m
//  zwpg-ios
//  开关按钮，继承自UIButton
//  Created by nijino on 13-4-27.
//  Copyright (c) 2013年 nijino. All rights reserved.
//  QQ:20118368
//  http://www.nijino.cn

#import "ToggleButton.h"

@implementation ToggleButton

- (void)setSelected:(BOOL)selected {
    if (selected) {
         [self setImage:self.onImage forState:UIControlStateNormal];
        [UIView animateWithDuration:0.5 delay:0.00 usingSpringWithDamping:0.7 initialSpringVelocity:5 options:UIViewAnimationOptionAllowUserInteraction animations:^{
            self.transform = CGAffineTransformMakeRotation(M_PI_2);
            
        } completion:^(BOOL finished) {
            
        }];
    }else {
         [self setImage:self.offImage forState:UIControlStateNormal];
        [UIView animateWithDuration:0.5 delay:0.00 usingSpringWithDamping:0.7 initialSpringVelocity:5 options:UIViewAnimationOptionAllowUserInteraction animations:^{
            self.transform = CGAffineTransformMakeRotation(0);
            
        } completion:^(BOOL finished) {
        }];
    }
}

+ (instancetype)buttonWithOnImage:(UIImage *)onImage
               offImage:(UIImage *)offImage
       highlightedImage:(UIImage *)highlightedImage{
    ToggleButton *button = [self buttonWithType:UIButtonTypeCustom];
    button.onImage = onImage;
    button.offImage = offImage;
    [button setImage:offImage forState:UIControlStateNormal];
    [button setImage:onImage forState:UIControlStateSelected];
    button.toggleEnabled = YES;
    button.userInteractionEnabled = YES;
    return button;
}


- (instancetype)initWithCoder:(NSCoder *)aDecoder{
    self = [super initWithCoder:aDecoder];
    if (self) {
        self.toggleEnabled = YES;

        [self setBackgroundImage:nil forState:UIControlStateSelected];

        [self setImage:nil forState:UIControlStateSelected];

    }
    return self;
}


#pragma mark - Toggle Support
- (BOOL)toggle{
    self.selected = !self.selected;
    return self.selected;
}

//Detect a touchUpInside event and perform toggle if toggleEnabled = YES
- (void)endTrackingWithTouch:(UITouch *)touch withEvent:(UIEvent *)event{
    [super endTrackingWithTouch:touch withEvent:event];
    if (self.touchInside && self.toggleEnabled) {
        [self toggle];
    }
}

@end

File: PracticeChinese/Module/BeginnerCourse/views/ToggleButton.swift
//
//  ToggleButton.swift
//  ChineseDev
//
//  Created by Temp on 2018/7/10.
//  Copyright © 2018年 msra. All rights reserved.
//

//import UIKit
//
class ToggleButton: UIButton {
    var toggleEnabled = false
    //开关是否有效
    var onImage: UIImage?
    //"开"状态图片
    var offImage: UIImage?
    //"关"状态图片
    var onBackgroundImage: UIImage?
    //"开"状态背景图片
    var offBackgroundImage: UIImage?
    //"关"状态背景图片
    var onTitle = ""
    //"开"状态标题
    var offTitle = ""
    //"关"状态标题
    var onTitleColor: UIColor?
    //"开"状态标题颜色
    var offTitleColor: UIColor?
    
    //"关"状态标题颜色
    //代码生成开关按钮类方法
    override init(frame: CGRect) {
        super.init(frame: frame)
        self.setSubviews()
    }
    
    func setSubviews() {
        self.onImage = UIImage(named: "in_recording")
        self.offImage = UIImage(named: "start_playing")
        self.setImage(offImage, for: .normal)
        self.setImage(onImage, for: .selected)
        self.toggleEnabled = true
        self.isUserInteractionEnabled = true
    }

    override var isSelected: Bool {
        didSet {
            /*
             if isSelected {
             UIView.animate(withDuration: 0.5, delay: 0.00, usingSpringWithDamping: 0.7, initialSpringVelocity: 5, options: .allowUserInteraction, animations: {
             self.transform = CGAffineTransform(rotationAngle: CGFloat(Double.pi / 2))
             }) { finished in
             }
             }else {
             UIView.animate(withDuration: 0.5, delay: 0.00, usingSpringWithDamping: 0.7, initialSpringVelocity: 5, options: .allowUserInteraction, animations: {
             self.transform = CGAffineTransform(rotationAngle: 0)
             }) { finished in
             }
             }
             */
        }
    }

    //  Converted to Swift 4 by Swiftify v4.1.6759 - https://objectivec2swift.com/
    required init?(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder)
        
        toggleEnabled = true
        setBackgroundImage(nil, for: .selected)
        setImage(nil, for: .selected)
        
    }
    
    // MARK: - Toggle Support
    func toggle() {
        self.isSelected = !isSelected
    }
    
    //Detect a touchUpInside event and perform toggle if toggleEnabled = YES
    override func endTracking(_ touch: UITouch?, with event: UIEvent?) {
        super.endTracking(touch, with: event)
//        if isTouchInside && toggleEnabled {
//            toggle()
//        }
    }


}
